Fixes #2267 - PXE templates may be re-rendered for selected hosts

xprazak2 opened the pull request:
Can be accessed from hosts index page

Commit:
Fixes #2267 - general rebuild of TFTP, DNS, DHCP

dLobatog commented:
@xprazak2 Tests failing, check that out :smile:

dLobatog commented on app/controllers/hosts_controller.rb:
```diff
@@ -421,6 +422,15 @@ def multiple_destroy
   def multiple_build
   end
 
+  def rebuild_tftp
```
Why this empty method?

xprazak2 replied:
It has similar purpose to multiple_build, it only shows a table of selected hosts and prompts user for confirmation.

dLobatog commented:
@xprazak2 Seems pretty solid at first glance :smile: I haven't tested or anything yet though. Two things that stand out:
1. Lack of tests
2. Builds encompass more than TFTP these days (the Redmine issue is old), why not rebuilding in general vs just regenerating TFTP templates?

Thanks for the patch!

xprazak2 commented:
TFTP re-rendering was related to another issue, that was the main reason I started working on this. I will go for general rebuild.

domcleal commented on test/functional/hosts_controller_test.rb:
```diff
@@ -311,6 +311,16 @@ def setup_user_and_host(operation, filter = nil, &block)
+  def test_multiple_rebuild_tftp
+    h = FactoryGirl.create(:host, :interfaces => [ FactoryGirl.build(:nic_primary_and_provision) ])
+    refute_nil h
+    refute_nil h.primary_interface
```
I don't think these assertions are helpful, just remove them.

domcleal commented on test/functional/hosts_controller_test.rb:
```diff
@@ -311,6 +311,16 @@ def setup_user_and_host(operation, filter = nil, &block)
+    refute_nil h
+    refute_nil h.primary_interface
+
+    post :submit_rebuild_tftp, { :host_ids => [h.id] }, set_session_user
```
It'd be helpful to have an expectation somewhere, e.g. you expect that setTFTP (or whatever) is called, or some proxy API etc.

domcleal commented:
> I will go for general rebuild.

That sounds very useful, thanks.  As part of this PR or separately?

domcleal commented on app/controllers/hosts_controller.rb:
```diff
@@ -421,6 +422,15 @@ def multiple_destroy
+  end
+
+  def submit_rebuild_tftp
+    @hosts.map(&:interfaces).flatten.select(&:provision).map { |nic| nic.send(:setTFTP) if nic.tftp? }
```
1. `@hosts` should be filtered somewhere to only be managed hosts, either in the selection list or perhaps here (though it seems a bit late).
2. I'd also suggest using host.provision_interface or simply `host.interfaces.where(:provision => true)` instead of filtering in-memory.
3. Maybe setTFTP should be a public method, to avoid send?

domcleal commented:
We might need a corresponding API, so this doesn't end up as a UI-only feature.

xprazak2 commented:
> As part of this PR or separately?

I guess most of this will be changed anyway, so probably as a part of this PR if nobody objects.

domcleal commented:
Suits me, thanks!

xprazak2 commented:
@domcleal, @eLobato: I've just found out, general rebuilds may already exist since [this feature](https://github.com/theforeman/foreman/blob/develop/app/controllers/hosts_controller.rb#L424-L437) rebuilds the whole hosts. I can add API if we want a feature that regenerates only TFTP. Your thoughts?

domcleal commented:
I thought he meant the issue with rebuilding of DNS and DHCP records etc, and not just the build flag.  

I think DHCP might have been fixed recently to verify the record exists when a host is saved, but it's a problem for folks when data is lost from the external server and they have to recreate it somehow.

xprazak2 commented:
DNS and DHCP records are now added to the external server when they are missing, we also have API and improved tests.

dLobatog commented on test/functional/api/v2/hosts_controller_test.rb:
```diff
@@ -301,6 +301,22 @@ def fact_json
+    host = FactoryGirl.create(:host)
+    post :rebuild_config, { :id => host.to_param }, set_session_user
+    assert_response :success
+    Host.any_instance.unstub(:rebuild_config)
```
No need to unstub at the end of the method as far as I know, the method won't be stubbed outside of this test's context.

xprazak2 commented:
[test]

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -203,6 +203,21 @@ def facts
         render_message(e.to_s, :status => :unprocessable_entity)
       end
 
+      api :POST, "/hosts/:id/rebuild_config", N_("Rebuild TFTP, DNS DHCP configuration")
```
a) Comment for throughout really, we may want to add a word into the name describe what sort of config, perhaps something like "rebuild provisioning config", "rebuild orchestrated config" or similar?
b) "TFTP, DNS and DHCP"

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -203,6 +203,21 @@ def facts
+        if failures.empty?
+          render :json => {:message => _("Configuration successfully rebuilt.")}, :status => :ok
+        else
+          render :json => {:message => _("Configuration rebuild failed for: %s.") % (failures.join(", "))}, :status => :unprocessable_entity
```
I'm a bit worried about the level of error reporting here, should we pass/raise full exceptions back to the client if they failed?

xprazak2 replied:
We are not passing full exceptions when there is a fail, we only provide a client with a message what failed like [this](http://i.imgur.com/gfYzsHG.png). I did not think we had to inform there was a Net::Error, but it certainly can be done. Would you like me to?

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -203,6 +203,21 @@ def facts
+          failures << key.to_s.upcase unless value
+        end
+        if failures.empty?
+          render :json => {:message => _("Configuration successfully rebuilt.")}, :status => :ok
```
Use render_message here.

domcleal commented on app/controllers/hosts_controller.rb:
```diff
@@ -414,6 +415,34 @@ def multiple_destroy
+  end
+
+  def submit_rebuild_config
+    all_fails = {:tftp => [], :dns => [], :dhcp => []}
```
Avoid hardcoding the orchestration types into the UI, I'd suggest filling this in on the fly during the loop below.

domcleal commented on app/controllers/hosts_controller.rb:
```diff
@@ -414,6 +415,34 @@ def multiple_destroy
+    message = ''
+    all_fails.each_pair do |key, values|
+      unless values.empty?
+        message << (_("%{r} rebuild failed for hosts: %{s}. ") % { :s => values.join(', '), :r => key.to_s.upcase})
```
a) Use more meaningful variables ('r', 's') in translated strings as translators need to understand the context.
b) For hostnames you may want to use `values.to_sentence` which gives a localised string, such as "foo, bar and baz".
c) Remove the whitespace at the end of the string - if it must be there, put it outside of the translation.

domcleal commented on app/models/host/managed.rb:
```diff
@@ -239,6 +239,20 @@ def built(installed = true)
+    nic = interfaces.select(&:provision).first
+    begin
+      result[:tftp] = nic.setTFTP if nic.tftp?
+    rescue
```
Where you rescue exceptions isn't consistent - DNS and DHCP are both rescued inside the rebuild methods, TFTP is rescued here.  I'm not sure if they should be rescued or not though, see above.  If they should be, I'd create a rebuild_tftp method and do it there for consistency.

domcleal commented on app/models/host/managed.rb:
```diff
@@ -239,6 +239,20 @@ def built(installed = true)
 
+  # rebuilds TFTP, DHCP, DNS for the host
+  def rebuild_config
+    result = {:tftp => false, :dhcp => false, :dns => false}
```
Any chance we can move this into the orchestration concerns somehow?  A bit like how they queue methods already, we could have a generic way to rebuild orchestration entries.

xprazak2 commented:
I moved `rebuild_config` from managed host into new `Orchestration::General` concern because it did not really fit into any of the existing. I also tried to make it as general as possible.

> I'm a bit worried about the level of error reporting here, should we pass/raise full exceptions back to the client if they failed?

I kept the messages in API controller simple and did not pass the full exceptions, because I was not sure you wanted me to. Client gets [this](http://i.imgur.com/gfYzsHG.png). If I should propagate the exceptions just let me know.

domcleal commented:
> I moved rebuild_config from managed host into new Orchestration::General concern because it did not really fit into any of the existing. I also tried to make it as general as possible.

Sorry, I think you may have missed my point a little.  Let me explain - normally an orchestration concern such as Orchestration::DHCP registers itself into the main class by using Rails callbacks (after_validation etc), so I was wondering if these concerns should do the same for the re-registration somehow, so we could just iterate over the list of orchestration concerns (all, or those registered for rebuilding support) and call rebuild_*.  It may be a little complex though.

Two reasons for this really: a) to possibly allow orchestration actions from plugins to add themselves to this, b) keep the logic of which orchestrations are enabled/disabled for a given host (remember, hosts come in many types, not all are unattended + managed) more flexible.

The other aspect of this is that currently the list of orchestration methods (tftp, dhcp, dns) is hardcoded into the response handling of the API and UI, which should be made generic.  (You already have a list of responses per type from the inner methods, reuse the keys.)

> > I'm a bit worried about the level of error reporting here, should we pass/raise full exceptions back to the client if they failed?
> 
> I kept the messages in API controller simple and did not pass the full exceptions, because I was not sure you wanted me to. Client gets this. If I should propagate the exceptions just let me know.

Yeah, we definitely need to do better somehow - since if it fails, how are you going to debug it?  At worst you could use Foreman::Logging.exception to ensure the logs of the error are kept at least.  That might be enough, but I think you should consider the pros/cons of letting errors bubble up.

xprazak2 commented:
I guess I missed your point by more than just a little :-) I will look into it, will be back

xprazak2 commented:
I think I've got it, more or less ;-). This time, there is no hardcoded list of orchestration methods. We check all the Orchestration::\* modules mixed into nic and call rebuild_\* method if the module has it, so it should be available for plugins. I also improved error handling - I decided to go mostly for logging because when user rebuilds multiple hosts and these have multiple errors, I do not think we could show reasonable details in a flash. Two questions on my mind, mostly related to plugins:

1) Naming convention of orchestration modules - will plugins use Orchestration::\* or ForemanDocker::Orchestration::*? Currently, first option is supported.

2) Some orchestration modules are already in host class and it may make more sense for plugins to add orchestration modules directly into the host instead of nic. I took the YAGNI approach, only modules in nic/managed get picked up and included in config rebuild. Do we want to support modules mixed into host?

domcleal commented:
> 1) Naming convention of orchestration modules - will plugins use Orchestration::\* or ForemanDocker::Orchestration::*? Currently, first option is supported.

I'd expect them to use the latter, since forcing people to use core namespaces rather than their own isn't great.  Perhaps in Orchestration you could have an array of rebuild methods, then in the orchestration concerns, in the `included` block, you'd do something like `register_rebuild(:rebuild_tftp, N_('TFTP'))` which could add the rebuild method into the array?

You can then simply iterate over the array when you need it.  Registering should avoid needing to assume something about the other orchestration concerns.

> Do we want to support modules mixed into host?

Yeah, might as well do both I'd say, since you already have `orchestration_methods` available in both host and NIC - it's an easy change to support host too, just iterate again in recreate_config.

domcleal commented on app/models/host/managed.rb:
```diff
@@ -848,6 +848,18 @@ def setup_clone
+  # returns  : Hash with 'true' if rebuild was a success for a given key (Example: {:tftp => true, :dhcp => false})
+  def recreate_config
+    result = {}
+    nic = interfaces.select(&:provision).first
```
You'd probably want to do _all_ NICs, because otherwise things such as DHCP configs on secondary NICs won't get recreated.

domcleal commented on app/models/concerns/orchestration/tftp.rb:
```diff
@@ -19,6 +19,21 @@ def tftp
+        true
+      end
+    rescue => e
+      logger.error "Failed to rebuild TFTP record for #{host.name}"
```
`Foreman::Logging.exception` can do all of this backtrace logging for you.

domcleal commented on config/initializers/monkeys.rb:
```diff
@@ -0,0 +1 @@
+require 'monkeys/symbol'
```
lib/core_extensions.rb is the place to put monkey patches, but you might see in my register_rebuild suggestion that I also included `N_('TFTP')` as an example.  You might be able to keep both the method name and a user-friendly name registered.

xprazak2 commented:
We now support modules mixed into both host and nic. All assumptions about orchestration modules are removed thanks to `register_rebuild` method as suggested above.

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -203,6 +203,21 @@ def facts
         render_message(e.to_s, :status => :unprocessable_entity)
       end
 
+      api :POST, "/hosts/:id/rebuild_config", N_("Rebuild orchestration config (TFTP, DNS and DHCP)")
```
I'd suggest leaving out the list of TFTP etc. here.

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -203,6 +203,21 @@ def facts
+      def rebuild_config
+        result = @host.recreate_config
+        failures = []
+        result.each_pair do |key, value|
```
A little more Ruby-ish might be to do something like: `result.reject { |k,v| v.blank? }.keys.map(&:to_s)`

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -203,6 +203,21 @@ def facts
+        if failures.empty?
+          render_message _("Configuration successfully rebuilt."), :status => :ok
+        else
+          render_message (_("Configuration rebuild failed for: %s.") % failures.join(", ")), :status => :unprocessable_entity
```
Use .to_sentence rather than join with commas.

domcleal commented on app/controllers/hosts_controller.rb:
```diff
@@ -416,6 +417,34 @@ def multiple_destroy
+    @hosts.each do |host|
+      result = host.recreate_config
+      result.each_pair do |k, v|
+        all_fails[k] = all_fails[k] || []
```
`all_fails[k] ||= []`

domcleal commented on app/controllers/hosts_controller.rb:
```diff
@@ -416,6 +417,34 @@ def multiple_destroy
+    message = ''
+    all_fails.each_pair do |key, values|
+      unless values.empty?
+        message << ((_("%{config_type} rebuild failed for hosts: %{host_names}.") % {:config_type => key.to_s.upcase, :host_names => values.to_sentence})) + " "
```
Use n_() here to support singular ("host") and plural ("hosts"), see http://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers

domcleal replied:
I don't think you need .upcase here, if the key is actually the "pretty name".  If it is the pretty name then I think you can do `_(key)` to get the localised version.

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -203,6 +203,21 @@ def facts
+        result = @host.recreate_config
+        failures = []
+        result.each_pair do |key, value|
+          failures << key.to_s unless value
```
Ditto what I said for the UI, but is `key` here the pretty name?  If so, call `_(key)` so we get the translated version.

domcleal commented on app/models/concerns/orchestration/dhcp.rb:
```diff
@@ -19,6 +20,20 @@ def dhcp_record
+      if dhcp?
+        set_dhcp
+      else
+        logger.warn "DHCP not supported for #{host.name}, #{ip}, skipping orchestration rebuild"
```
Use debug or info I think here, since this is a potentially normal condition.

domcleal commented on app/models/concerns/orchestration/dns.rb:
```diff
@@ -18,6 +19,20 @@ def reverse_dns?
+      if dns? && reverse_dns?
+        recreate_a_record && recreate_ptr_record
+      else
+        logger.warn "DNS not supported for #{host.name}, #{ip}, skipping orchestration rebuild"
```
ditto about logging level

domcleal commented on app/models/concerns/orchestration/dns.rb:
```diff
@@ -18,6 +19,20 @@ def reverse_dns?
 
+  def rebuild_dns
+    begin
+      if dns? && reverse_dns?
```
It's quite possible to have one and not the other, these should both be run, not only when they are both true.

domcleal commented on app/models/concerns/orchestration/tftp.rb:
```diff
@@ -19,6 +20,20 @@ def tftp
+      if tftp?
+        setTFTP
+      else
+        logger.warn "TFTP not supported for #{host.name}, #{ip}, skipping orchestration rebuild"
```
ditto about log level

domcleal commented on app/models/host/managed.rb:
```diff
@@ -848,6 +848,26 @@ def setup_clone
+    end
+
+    self.class.rebuild_methods.map do |method, pretty_name|
+      fail "The name '#{pretty_name}' is already taken by some other orchestration module, please choose another name" if result[pretty_name]
```
This could be problematic, as it will fail at runtime for the user while the other "fail" is at load time (so we'd spot it during tests).  I think it should raise a Foreman::Exception here, and make the message more user-friendly rather than developer-centric.

domcleal commented on app/services/foreman/access_permissions.rb:
```diff
@@ -347,7 +347,8 @@
                                                :disassociate, :multiple_disassociate, :update_multiple_disassociate,
-                                               :select_multiple_location, :update_multiple_location].push(*ajax_actions),
+                                               :select_multiple_location, :update_multiple_location, :rebuild_config,
+                                               :submit_rebuild_config].push(*ajax_actions),
```
There's an inconsistency here between the UI and API - the API permission is on build_hosts, but the UI permission is on edit_hosts.  I'd suggest perhaps moving the UI permission to build_hosts.

domcleal commented on test/functional/api/v2/hosts_controller_test.rb:
```diff
@@ -301,6 +301,20 @@ def fact_json
   end
 
+  def test_rebuild_config_optimistic
+    Host.any_instance.stubs(:recreate_config).returns({ "TFTP" => true, "DNS" => true, "DHCP" => true })
```
Use .expects rather than stubs so we ensure it was called.

domcleal commented on test/functional/api/v2/hosts_controller_test.rb:
```diff
@@ -301,6 +301,20 @@ def fact_json
+  end
+
+  def test_rebuild_config_pessimistic
+    Host.any_instance.stubs(:recreate_config).returns({ "TFTP" => false, "DNS" => false, "DHCP" => false })
```
stubs => expects

domcleal commented on test/functional/hosts_controller_test.rb:
```diff
@@ -881,6 +881,28 @@ class Host::Valid < Host::Managed; end
 
+  def test_submit_multiple_rebuild_config_optimistic
+    @request.env['HTTP_REFERER'] = hosts_path
+    Host.any_instance.stubs(:recreate_config).returns({"TFTP" => true, "DHCP" => true, "DNS" => true})
```
stubs/expects

domcleal commented on test/functional/hosts_controller_test.rb:
```diff
@@ -881,6 +881,28 @@ class Host::Valid < Host::Managed; end
+
+  def test_submit_multiple_rebuild_config_pessimistic
+    @request.env['HTTP_REFERER'] = hosts_path
+    Host.any_instance.stubs(:recreate_config).returns({"TFTP" => false, "DHCP" => false, "DNS" => false})
```
stubs/expects

domcleal commented:
1) HostTest should also have some tests for the logic in recreate_config around overwriting results only when they're false etc., plus test for the failure when a host orchestration has the same name as a NIC orchestration.

2) Just check that the pretty name is making it through, and call `_()` closer to the UI when it can be translated.

Looking good!

xprazak2 commented:
I added more tests for `recreate_config` into HostTest and replaced `stubs` with `expects` in the already existing ones. I removed the inconsistency in permissions, changed logging levels + made some minor improvements in style. The pretty_names should make it all the way to the user as far as I can tell.

domcleal commented on app/models/concerns/orchestration/dns.rb:
```diff
@@ -18,6 +19,20 @@ def reverse_dns?
+  def rebuild_dns
+    begin
+      if dns? || reverse_dns?
+        recreate_a_record && recreate_ptr_record
```
You should only call the recreate_X_record if that type of DNS is enabled, so don't call recreate_ptr_record is dns? is true and reverse_dns? is false.

I think using `&&` here could be a problem too, as it would skip rebuild_ptr_record if rebuild_a_record returned a falsey value.

domcleal commented on app/models/concerns/orchestration/dhcp.rb:
```diff
@@ -19,6 +20,20 @@ def dhcp_record
+        logger.info "DHCP not supported for #{host.name}, #{ip}, skipping orchestration rebuild"
+        true
+      end
+    rescue => e
```
Nitpick, can you move this begin/rescue block around the set_\* method specifically, ditto in all other orchestration types.  It avoids the danger of us catching an exception in surrounding code and misreporting it.

domcleal commented on app/models/host/managed.rb:
```diff
@@ -848,6 +848,27 @@ def setup_clone
+    end
+
+    self.class.rebuild_methods.map do |method, pretty_name|
+      raise ::Foreman::Exception.new(_("There are orchestration modules with methods for configuration rebuild
```
Use N_() in combination with Foreman::Exception - we pass in the English string in order to generate a consistent error code, then translate it internally.  Also keep the string on one line as the i18n string extractor will think the \n is inside the string.

```
::Foreman::Exception.new(N_("There are.. %s"), pretty_name)
```

see http://projects.theforeman.org/projects/foreman/wiki/Translating#Exceptions

domcleal commented on app/controllers/hosts_controller.rb:
```diff
@@ -578,7 +609,7 @@ def action_permission
           'update_multiple_organization', 'select_multiple_organization',
           'update_multiple_location', 'select_multiple_location',
-          'disassociate', 'update_multiple_disassociate', 'multiple_disassociate'
+          'disassociate', 'update_multiple_disassociate', 'multiple_disassociate', 'rebuild_config', 'submit_rebuild_config'
```
These need moving into the "build" block around line 598.  Where they are means that the user will need edit_hosts permission to use the build actions, when the other change (to access_permissions) assumes build_hosts instead.

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -247,6 +259,8 @@ def action_permission
           when 'vm_compute_attributes'
             :view
+          when 'rebuild_config'
+            :edit
```
:build

domcleal commented on app/models/concerns/orchestration/dhcp.rb:
```diff
@@ -19,6 +20,20 @@ def dhcp_record
+      if dhcp?
+        set_dhcp
+      else
+        logger.info "DHCP not supported for #{host.name}, #{ip}, skipping orchestration rebuild"
```
Nitpick, use #{self} rather than host.name here, as DHCP might be on a non-primary interface, so showing the NIC (self) it's on will be more precise.

ditto on others please.

domcleal commented on app/models/concerns/orchestration/dhcp.rb:
```diff
@@ -19,6 +20,20 @@ def dhcp_record
+  def rebuild_dhcp
+    begin
+      if dhcp?
+        set_dhcp
```
Should we also run removal methods (e.g. del_dhcp) in these methods? Perhaps throwing away the result?  I was just trying to test it with a host that already had its entries, and of course it fails.  I'm not sure if that'd be expected, a user would instead probably expect it to destroy/create.

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -203,6 +203,18 @@ def facts
         render_message(e.to_s, :status => :unprocessable_entity)
       end
 
+      api :POST, "/hosts/:id/rebuild_config", N_("Rebuild orchestration config")
```
use PUT for this API, it's more consistent with other actions

xprazak2 commented:
I changed POST into PUT in API and permissions to `:build` to maintain consistency. We now delete already existing records before setting new ones, so there should be no more errors.

domcleal commented on app/views/hosts/rebuild_config.html.erb:
```diff
@@ -0,0 +1,5 @@
+<%= render 'selected_hosts', :hosts => @hosts %>
+
+<%= form_tag submit_rebuild_config_hosts_path(:host_ids => params[:host_ids]), :onsubmit => "resetSelection()" do %>
+  <small><%= _('Rebuild configuration for selected hosts') %></small>
```
It'd be worth saying "Rebuild orchestration configuration" here to match the API, so it's clearer what it's for.  There's limited space on the dropdown menu to say precisely what config it is, but here's better.

domcleal commented:
Looking good - two comments:

1) Could you add some unit tests for the rebuild methods to test/unit/orchestration/?  Check it calls the right methods when DNS/DHCP etc are/aren't enabled.
2) The warning on the templates page made in a377f776 could be updated to point users to this new feature.

xprazak2 commented:
I've added tests for rebuild methods and a link from templates page which should point users in the right direction.

domcleal commented on test/unit/orchestration/dhcp_test.rb:
```diff
@@ -195,4 +195,24 @@ def teardown
+    h = FactoryGirl.create(:host, :with_dhcp_orchestration)
+    Nic::Managed.any_instance.expects(:del_dhcp)
+    Nic::Managed.any_instance.expects(:set_dhcp).returns(true)
+    assert h.recreate_config['DHCP']
```
Nitpick - in these tests just call your new rebuild_dhcp type method.

domcleal commented on test/unit/orchestration/dhcp_test.rb:
```diff
@@ -195,4 +195,24 @@ def teardown
+    assert h.recreate_config['DHCP']
+  end
+
+  test "should skip dhcp rebuild" do
```
You may want to add an `expects(:set_dhcp).never` type check.

domcleal commented on test/unit/orchestration/dns_test.rb:
```diff
@@ -82,4 +82,41 @@ def test_bmc_should_have_valid_dns_records
+    h = FactoryGirl.create(:host, :with_dns_orchestration)
+    Nic::Managed.any_instance.expects(:del_dns_a_record_safe)
+    Nic::Managed.any_instance.expects(:del_dns_ptr_record_safe)
+    Nic::Managed.any_instance.expects(:recreate_a_record).returns(true)
```
I think these should all be stubbed one level lower, so the recreate_\* and *_safe methods get some test coverage.  Stub `:set_dns_a_record` and `:del_dns_a_record` instead.

domcleal commented on app/views/templates/_form.html.erb:
```diff
@@ -38,8 +38,9 @@
 
 
       <% if pxe_with_building_hosts?(@template) -%>
+        <% warning_text = (_("Warning: The template is associated to at least one host in build mode. To apply the change, disable and enable build mode on hosts to update the live templates or choose to %s their configuration from 'Select Action' menu") % link_to(_('recreate'), hosts_path(:host_ids => building_hosts(@template).map(&:id)))).html_safe %>
```
This link loads the hosts index, but unless you happen to have the cookie already then the select action dropdown isn't shown, and the hosts aren't selected.

xprazak2 commented:
I stubbed the methods one level lower and cookies should be working at all times.

sodabrew commented:
I have a question: will this also update the DHCP entries for other NICs, e.g. an IPMI interface?

xprazak2 commented:
@sodabrew: It will update DHCP entries of all the host's NICs.

domcleal commented on app/helpers/provisioning_templates_helper.rb:
```diff
@@ -67,7 +67,17 @@ def pxe_with_building_hosts?(template)
+
+  def link_to_selected_hosts(text, host_ids)
+    cookies.delete(:'_ForemanSelectedhosts') if cookies[:'_ForemanSelectedhosts']
+    cookies[:'_ForemanSelectedhosts'] = { :value => JSON.generate(host_ids), :secure => true }
```
Couple of things here:

a) I don't think we should change the user's selected cookie just by viewing the provisioning template page, this could be very surprising and even dangerous.  I'd suggest either not trying to implement this functionality, or do it via the search function - such as adding a new (hidden) search term to hosts for `uses_template_id = 6` or something that's easy to link to.
b) secure => true only works when the server's HTTPS.

domcleal commented on app/views/templates/_form.html.erb:
```diff
@@ -37,8 +37,9 @@
 
 
       <% if pxe_with_building_hosts?(@template) -%>
+        <% warning_text = (_("Warning: The template is associated to at least one host in build mode. To apply the change, disable and enable build mode on hosts to update the live templates or choose to %s their configuration from 'Select Action' menu") % link_to_selected_hosts(_('recreate'), building_hosts(@template).map(&:id))).html_safe %>
```
Probably obsolete given my comment above, but prefer .pluck(:id) rather than .map(&:id) as it'll run in SQL instead.

xprazak2 commented:
You are right, we definitely should not modify the cookie when viewing the template. 
Changing the user's selection can be unexpected, but the table operates on the cookie so I guess redirect with unselected hosts might be a feasible compromise.

domcleal commented on app/helpers/provisioning_templates_helper.rb:
```diff
@@ -67,7 +67,15 @@ def pxe_with_building_hosts?(template)
+  end
+
+  def building_hosts_path(template)
+    hosts_path(:search =>  "name = " << building_hosts(template).pluck(:name).join(" or ") )
```
Instead of listing every host, you could construct a shorter search query which may help when you have _many_ hosts, which would hit the URI length limit, e.g. `build = true and (os_id = 1 or os_id = 2)`

domcleal commented:
> Changing the user's selection can be unexpected, but the table operates on the cookie so I guess redirect with unselected hosts might be a feasible compromise.

Yep, works for me, it's easy to select all if you then wish to, or just the hosts you're interested in.

xprazak2 commented:
I changed the query to use os_id

domcleal commented:
Merged as 2fba6ad, thanks @xprazak2!
